### DISCUSSION TITLE
Why is Signal not supported?

### DISCUSSION THREAD

**Discussion by barracuda156** [d99kris/nchat#317] [Q&A]
@d99kris I have seen a note in README, but wanna ask what is the reason for it: is it unfeasible, or you do not want it, or something else?
There are no portable Signal clients atm, AFAIK, it would be awesome to have one.

**Comment by pdaian** [comment]
https://github.com/boxdot/gurk-rs works well and has a single-binary package if that's what you mean by "portable"

but agree it would be cool to merge the two somehow

**Reply by barracuda156** [reply]
> https://github.com/boxdot/gurk-rs works well and has a single-binary package if that's what you mean by "portable"

@pdaian Thank you for a reference, but that won’t work on systems without llvm and rust, so IMO it is not considered portable. (In particular, it does not work for me.)
Once a) `mrustc` reaches a functional state and b) rust is finally capable of using gcc backend, that may change, but until then it is both impossible and undesirable (dependencies are insanely numerous and heavy).

**Reply by d99kris** [reply]
> https://github.com/boxdot/gurk-rs works well and has a single-binary package if that's what you mean by "portable"
> 
> but agree it would be cool to merge the two somehow

Thanks for sharing @pdaian 👍

**Comment by d99kris** [comment]
Hi @barracuda156 - if I recall correctly I think there were three main reasons:
- There was no open-source client library implementation that was simple to integrate into nchat (I think I did manage to build the official Java client library and link it into nchat, but code-wise it was a bit messy).
- The official stance from the Signal project was [not really supportive of third-party clients](https://github.com/LibreSignal/LibreSignal/issues/37#issuecomment-217211165)
- I don't use Signal myself

I think it could be reconsidered if someone was willing to implement and help maintain that protocol part, but ideally I'd prefer some type of plugin model where the protocol is implemented in a separate project.

One possibility could maybe be to add support for using `libpurple` once its v3 is ready, and indirectly support the protocols it supports. Last time I heard they had only Jabber support in v3, but given their rangle of plugins in v2, I would expect v3 to support more than Jabber eventually. However, I have not really checked the v3 API, so I'm not sure yet how feasible or practicle it would be to use.

**Comment by barracuda156** [comment]
@d99kris To avoid opening another issue for a small question, asking here: is adding some p2p protocol support possible? There are at least three: tor (used by Speek), i2p (used by i2pchat) and tox (used by toxic, uTox, qTox and tomato).

**Reply by d99kris** [reply]
While technically feasible, I think I'll need to leave it out of scope, at least for foreseable future. Unless someone is willing to contribute and maintain it (so basically same as my reply above regarding Signal).

If I had a lot more time it would be nice to add more protocols. But realistically, I don't think that will happen.

**Reply by barracuda156** [reply]
Got it. This is not my area of expertise, unfortunately. But I hope someone will get an interest to implement i2p and tor in a CLI messenger, since currently those got only Qt-based implementations, and AFAIK seldom if at all maintained. (I mean specifically for messaging, not i2p and tor as such, of course.) Tox has its CLI client.

**Comment by d99kris** [comment]
Please see #426 for an update on this, basic Signal functionality is available for beta-testing.
